core/unix/coredump.c

```diff
@@ -154,6 +194,178 @@ write_section_header(DR_PARAM_IN file_t elf_file,
     return os_write(elf_file, (void *)&shdr, sizeof(shdr)) == sizeof(shdr);
 }
 
+/*
+ * Copy dr_mcontext_t register values to user_regs_struct.
+ */
+static void
+mcontext_to_user_regs(DR_PARAM_IN dr_mcontext_t *mcontext,
+                      DR_PARAM_OUT struct user_regs_struct *regs)
+{
+#ifdef DR_HOST_NOT_TARGET
+    return;
+#elif defined(X86)
+    regs->rax = mcontext->rax;
+    regs->rcx = mcontext->rcx;
+    regs->rdx = mcontext->rdx;
+    regs->rbx = mcontext->rbx;
+    regs->rsp = mcontext->rsp;
+    regs->rbp = mcontext->rbp;
+    regs->rsi = mcontext->rsi;
+    regs->rdi = mcontext->rdi;
+    regs->r8 = mcontext->r8;
+    regs->r9 = mcontext->r9;
+    regs->r10 = mcontext->r10;
+    regs->r11 = mcontext->r11;
+    regs->r12 = mcontext->r12;
+    regs->r13 = mcontext->r13;
+    regs->r14 = mcontext->r14;
+    regs->r15 = mcontext->r15;
+    regs->rip = (uint64_t)mcontext->rip;
+    regs->eflags = mcontext->rflags;
+#elif defined(AARCH64)
+    regs->regs[0] = mcontext->r0;
+    regs->regs[1] = mcontext->r1;
+    regs->regs[2] = mcontext->r2;
+    regs->regs[3] = mcontext->r3;
+    regs->regs[4] = mcontext->r4;
+    regs->regs[5] = mcontext->r5;
+    regs->regs[6] = mcontext->r6;
+    regs->regs[7] = mcontext->r7;
+    regs->regs[8] = mcontext->r8;
+    regs->regs[9] = mcontext->r9;
+    regs->regs[10] = mcontext->r10;
+    regs->regs[11] = mcontext->r11;
+    regs->regs[12] = mcontext->r12;
+    regs->regs[13] = mcontext->r13;
+    regs->regs[14] = mcontext->r14;
+    regs->regs[15] = mcontext->r15;
+    regs->regs[16] = mcontext->r16;
+    regs->regs[17] = mcontext->r17;
+    regs->regs[18] = mcontext->r18;
+    regs->regs[19] = mcontext->r19;
+    regs->regs[20] = mcontext->r20;
+    regs->regs[21] = mcontext->r21;
+    regs->regs[22] = mcontext->r22;
+    regs->regs[23] = mcontext->r23;
+    regs->regs[24] = mcontext->r24;
+    regs->regs[25] = mcontext->r25;
+    regs->regs[26] = mcontext->r26;
+    regs->regs[27] = mcontext->r27;
+    regs->regs[28] = mcontext->r28;
+    regs->regs[29] = mcontext->r29;
+    regs->regs[30] = mcontext->r30;
+    regs->sp = mcontext->sp;
+    regs->pc = (uint64_t)mcontext->pc;
+#endif
+}
+
+/*
+ * Write prstatus structure to the file. Returns true if the note is written to
+ * the file, false otherwise.
+ */
+static bool
+write_prstatus_note(DR_PARAM_IN file_t elf_file)
+{
+    void *drcontext = dr_get_current_drcontext();
+    dr_mcontext_t mcontext;
+    mcontext.size = sizeof(mcontext);
+    mcontext.flags = DR_MC_ALL;
+    dr_get_mcontext(drcontext, &mcontext);
```

Comment:
This assume a client calling context (e.g., a clean call): this is not safe to be called from DR which avoids clean calls and generally has done a full switch back to DR state, and AFAIK there's no flag or anything telling this routine where it is outside of the client API routines where it's supported.

The os_dump_core* routines in general are called from DR's own code.  The same for os_dump_core_live() on Windows.  If this Linux os_dump_core_live() doesn't support DR calls that needs to be made clear.  I don't think we want to do that though: this could be a useful internal DR debugging feature parallel to the current fork-and-core-dump.

One way forward is to pass the mcontext all the way through: so the API routine dr_create_memory_dump() would be the one caling dr_get_mcontext_priv(); it's safe there as that's a client-only-called routine.

Reply:
I've updated the code to call dr_get_mcontext_priv() in dr_create_memory_dump().